### ISSUE TITLE
Apply VA_SCALE to JWST WCS objects

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#5434]
_Issue [JP-1762](https://jira.stsci.edu/browse/JP-1762) was created on JIRA by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse):_

Level 1b JWST processing computes velocity aberration effects on coordinate shifts and image scale, storing resulting values in the DVA_RA, DVA_DEC, and VA_SCALE keywords. Currently none of these values is used downstream to apply any corrections to the data files. VA_SCALE should be applied to WCS objects at some point in the calibration processing, so that the change in scale is accounted for, as is done for HST images using their VAFACTOR keyword.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=499098#comment-499098):_

[Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) Could you link the soccer document discussing the velocity aberration here. It's not in the meeting notes.

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=499175#comment-499175):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) sure, here are the details for the SOCCER document describing the detailed formulas for both corrections (ie the velocity aberration along the telescope boresight optical axis, as well as the differential velocity aberration for a given instrument off-axis):

JWST-STScI-001888, R. Henry, J, Isaacs, W, Kinzel, "JWST Velocity Aberration Correction", April 16, 2010, Rev A

There are details on how it's implemented for HST, as well as specific discussions in the context of JWST.

Let me know please whether or not you're able to find this in SOCCER successfully (sorry, I'm not sure how to post the direct link to the document in SOCCER).

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=501302#comment-501302):_

Linking this ticket to <https://jira.stsci.edu/browse/JP-244> since there's commonality, and to ensure that relevant changes are made in a consistent way.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=503081#comment-503081):_

Main points from our correspondence with Colin Cox:
 # "... the main correction is already applied. This can be up to a 20 arc-sec shift so you could easily miss your target if it wasn’t." - This is essentially a pointing correction.
 # "Then a scale correction should be adequate." Full relativistic "scale" correction is quite complicated but Colin showed that using a single scale factor produces very good results:

 
```java
My comparison covered a 200 arc-sec area corresponding roughly to a typical detector. Then the constant scale approximation is good to within micro arc-seconds. I tried 2000 arcsec and then found discrepancies of a few milli arc-sec. Still too small to concern us in pointing activities.```
Here is Colin's notebook for reference: [^Aberration.ipynb]

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=503631#comment-503631):_

[Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara)  for a somewhat more detailed comparison, could you take a look at producing values or plots similar to Figs 5 and 6 in the SOCCER document ([JWST-STScI-001888](https://soccer.stsci.edu/experience/#documents/view/321YXDC_0000803C200006Y/document/321YYD6_001ZKQBCH0004KF))? - especially for their worst-case, ie considering the maximum distance from FGS to the science instruments, and considering also the position in the ecliptic?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=503643#comment-503643):_

When I try to access the SOCCER document, I get:
```java
You do not have the necessary privileges to perform this action. (B1ZQMG:G6TKJV)
```
Could you please put it (if security allows) to a location accessible to me?

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=503657#comment-503657):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) - could you confirm please if you're able to access the SOCCER document, and if so, could you provide it to [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara)  please?

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=504013#comment-504013):_

I copied all documents related to VA that I know of to 

/user/dencheva/vel-aberration

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=504808#comment-504808):_

The VA scale factor needs to be set in at least one dataset for each mode in the RT data so that the new code runs.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=504826#comment-504826):_

??Level 1b JWST processing computes velocity aberration effects on coordinate shifts and image scale, storing resulting values in the DVA_RA, DVA_DEC, and VA_SCALE keywords. Currently none of these values is used downstream...??

*Constant scale* velocity aberration *correction* was implemented in #5602 The code relies on `va_scale` value from `meta.velocity_aberration.scale_factor`. However, it does not look like this value is actually computed because it seems like it is set to None by the time when `assign_wcs` is run for all datasets in the regression tests.

Where do we get the `vx, vy, vz` from (where in the `meta`) to run code that computes constant scale factor in <https://github.com/spacetelescope/jwst/blob/master/jwst/lib/set_velocity_aberration.py](https://github.com/spacetelescope/jwst/blob/master/jwst/lib/set_velocity_aberration.py?)> ? Is this the code that should be run in Level 1b?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=505094#comment-505094):_

The code for applying DVA correction can be seen in <https://github.com/spacetelescope/jwst/pull/5602].>

Code for correcting DVA consists of two distinct steps:
 1. create a transformation that represents DVA correction;
 2. chain this transformation to the last step in the transformations from detector to `V2-V3` coordinates.

The model that represents DVA fundamentally applies a constant scale transformation in a tangent plane perpendicular to `V1`. This is implemented in the following way:
 1. Convert spherical coordinates `v2, v3` to cartesian representation `(x,y,z)`. X is the axis perpendicular to the image.
 2. Rotate `(x,y,z)` coordinates by `v2_ref`, and `-v3_ref` angles in order to align the `X` axis with `V1`.
 3. Multiply (rotated) two coordinates `(y',z')` by VA_SCALE while leaving `x'` coordinate unchanged, effectively applying scale in a tangent plane.
 4. Convert back re-scaled cartesian coordinates `(x',y',z')` to spherical coordinates `(v2',v3')`.

These transformed `(v2',v3')` coordinates account for DVA using a constant scale transformation.

The core code for this transformation is in [jwst/assign_wcs/pointing.py](https://github.com/spacetelescope/jwst/pull/5602/files#diff-6971ad1a81686167406e12643b01a0cc4f0e21917a27497592fa6c63dce78991R241)

NOTE: This procedure relies on the VA scale provided in `datamodel.meta.velocity_aberration.scale_factor`

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=505195#comment-505195):_

In order to help reviewing the code for the core algorithm that corrects (v2, v3) angles for DVA, I am attaching a notebook that contains simplified version of the code without deep dependencies. Everything is self-contained except for numpy dependency.

[^vacorr_model.ipynb]

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=505196#comment-505196):_

And here I am attaching reproduction of the Fig 5 from  JWST-STScI-001888 obtained using similar parameters to those in the paper and using the constant DVA scale value computed with `jws.lib.set_velocity_aberration.aberration_scale()` implemented in the pipeline.

[^fig5.ipynb]

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=505244#comment-505244):_

"Reproduction" of Fig. 5. Image extracted from notebook fig5.ipynb:

!Fig5-reproduction.png!

**Comment by stscijgbot-jp** [comment]
_Comment by [Colin Cox](https://jira.stsci.edu/secure/ViewProfile.jspa?name=cox) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=505909#comment-505909):_

Applying the scale correction in the V2V3 plane does not require any conversions or rotating. The V2V3 plane is the tangent plane and since the scale is isotropic in this plane we can just do something like

V_corrected - Vref = scale*(V -Vref) where V is V2 or v3. This presumes the pointing direction is more or less towards the V2V3 reference point. 

I have not finished looking at the other software.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=506014#comment-506014):_

??Applying the scale correction in the V2V3 plane does not require any conversions or rotating. The V2V3 plane is the tangent plane ...??

I am confused. Are the V2,V3 angles computed by the gWCS pipeline spherical angles or not? Or do you mean that with an acceptable accuracy we can approximate a sphere with a plane? I.e., xtangent = Tan(v2) = v2 + v2^3/3 - ... and so with some accuracy xtangent = v2?

 

Otherwise it is not clear to me why in the formula on page 15 in JWST-STScI-001550 there are 5 rotations with two of them R2(v3) R3(-v2) (here, I think v2, v3 are interpreted as v2_ref and v3_ref - see quote below):

??We will also know for at least one (v2,v3) point what the corresponding absolute position (α,δ) is (this may be the (v2,v3)=(0,0) position corresponding to the V1 axis... [hence my conclusion that v2 in that equation is actually what we call v2_ref in our data structures]??

??First, we align the position (α,δ) with the V1 unit vector. This is achieved by the successive rotations R3(-v2) and R2(v3).??

If v2, v3 were already in the tangent plane, shouldn't we simply subtract v2_ref and v3_ref instead of rotating by those angles?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=506058#comment-506058):_

I did some testing on a NIRCAM image and indeed keeping linear terms results in (what I think are) negligible errors (~1e-9 arcsec). I did not do any checks for any other instruments but my gut feeling is that the linear approximation (xt=v2, yt=v3) should be good enough. I will update my correction code to implement this approximation. CC: [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva)

**Comment by stscijgbot-jp** [comment]
_Comment by [Colin Cox](https://jira.stsci.edu/secure/ViewProfile.jspa?name=cox) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=506075#comment-506075):_

We have always indulged in a certain amount of doublethink with respect to V2V3. For large variations such as calculating pointings or relating positions between different apertures the full 3-dimensional rotation should be done. For deviations of a few hundred arc-sec on a detector, the planar approximation is OK. This is what we use in generating distortion solutions and calculating guide star offsets. So I believe everything is consistent with this.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=506077#comment-506077):_

Realized it would be useful to paste my comments here, not just on the meeting minutes page.

I've gone digging around testing utc_to_tdb against Nops9 data. It looks like jwst_x/y/z and jwst_dx/dy/dz are all geocentric in units of km and km/s. For the barycentric velocity correction to the wavelength solution, there's a rather complicated set of computations that advances the spacecraft position forward/back along the velocity vector 500 seconds, computes new positions adding in geocentric-barycentric offsets with the JPL ephemeris, computes the projection of these vectors along the line of sight to the object, converts that into a time at speed c, takes the difference of the two points, then divides out c again to get a barycentric velocity of the spacecraft towards the target. I'm really not sure why this is so excessively complicated, but the upshot is that VELOSYS should be in m/s relative to the solar system barycenter.

I'd be really interested in knowing the reason behind this long computation (looks like Phil's); hopefully the Nops10 data looking at a point near the ecliptic will confirm the numbers soon.

For the VA correction, we should be aware that dx/dy/dz seem to not include the (dominant) component due to Earth's orbital velocity. It should be easy enough to add that in though, as astropy.coord.get_body_barycentric could be swapped out for get_body_barycentric_posvel which returns both position and velocity.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=506157#comment-506157):_

Using approximate instrument positions from <https://jwst-docs.stsci.edu/files/97976960/97976962/1/1596073034159/JWST+FOV.png> I estimate that maximum error when using linear approximation xt=v2, yt=v3 versus full rotation & projection will be ~2e-3 mas (or 5e-5 NIRCAM pixels). It still seems small but if there are more approximations like this involved then the accuracy loss may accumulate over the entire WCS pipeline. This may be important for those who are trying to achieve 1e-3 - 1e-4 pixels alignment accuracy.

This approximation that v2 and v3 are already tangent plane coordinates results in a 3x speed improvement in DVA correction.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=507303#comment-507303):_

For the record, the module that computes parameters for VA correction (scale and aberrated RA and Dec) has been almost completely re-designed - see <https://github.com/spacetelescope/jwst/pull/5666.>

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=510188#comment-510188):_

For testing this, we'll want to add in the appropriate VA_SCALE keywords to both imaging and spectral data in the jwst regression tests.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=510929#comment-510929):_

And it turns out that lots of our regression test data already had VA_SCALE keywords added as it came from OTB or SDP.  So we're good on this front.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=510934#comment-510934):_

In PR #5602 I added a new frame to the WCS pipeline named  'v2v3vacorr' which accounts for DVA effects using Colin Cox suggestion that v2 and v3 angles can be already considered to be in a tangent plane and so current implementation is to perform a simple scale transformation in the v2v3 "plane".

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1762?focusedCommentId=510936#comment-510936):_

Resolved via #5602
